lib/src/gainmapmath.cpp

```diff
@@ -937,4 +1113,70 @@ std::unique_ptr<uhdr_raw_image_ext_t> convert_raw_input_to_ycbcr(uhdr_raw_image_
   return dst;
 }
 
+static bool floatToUnsignedFractionImpl(float v, uint32_t maxNumerator, uint32_t* numerator,
+                                        uint32_t* denominator) {
+  if (std::isnan(v) || v < 0 || v > maxNumerator) {
+    return false;
+  }
+
+  // Maximum denominator: makes sure that the numerator is <= maxNumerator and the denominator
+  // is <= UINT32_MAX.
+  const uint64_t maxD = (v <= 1) ? UINT32_MAX : (uint64_t)floor(maxNumerator / v);
+
+  // Find the best approximation of v as a fraction using continued fractions, see
+  // https://en.wikipedia.org/wiki/Continued_fraction
+  *denominator = 1;
+  uint32_t previousD = 0;
+  float currentV = v - floor(v);
```

Comment:
I'd suggest to change all intermediate float results to double type for better precision.

IEEE single-precision float only has 23 bits of significand which is not enough to carry out continued fraction for numerator and denominators as 32-bit integers. To make matters worse, you have a while loop with up to 39 iterations, so errors will accumulate and become unacceptably large.